src/Collection/CollectionTrait.php

```diff
@@ -193,6 +193,54 @@ public function min($callback, $type = SORT_NUMERIC)
     /**
      * {@inheritDoc}
      */
+    public function avg($matcher = null)
+    {
+        $iterator = $this->unwrap();
+        $count = $iterator instanceof Countable ?
+            count($iterator) :
+            iterator_count($iterator);
+
+        if ($count === 0) {
+            return null;
+        }
+
+        return $this->sumOf($matcher) / $count;
```

Comment:
I would say that the only advantage of having this method in the core is so that really large iterators (and specially generators) can be used without consuming all the memory or throwing an exception.

This implementation will, for example, throw an exception if the internal iterator in the collection is a generators (a function calling `yield`)

a better implementation would calculate the result on the fly as it iterates on the result. This is one version:

```php
$result =
$this
    ->extract($matcher)
     ->reduce(function($acc, $current) {
         list($count, $sum) = $acc;
         return [$count + 1, $sum + $current]
    }, [0, 0]);
return $result[1] / $result[0];
```